rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+*   Performance tests: Run a curated set of models for a pre-selected number of steps. Measure steps per second, and for image models images per second.
+
+
+## Releases
+
+In the multi package environment, we aim to have the following with the releases:
+
+1.  Smaller packages: We expect to release multiple small packages, instead of the giant package we are building and releasing right now.
+1.  Faster releases: We would like the smaller packages to be able to release much faster. With the ability to pin each dependency to a known good version, we would like to be able to compartmentalize issues, and only hold certain components back when only certain code in TF has issues.
+1.  Independent releases: With compatibility checks, we would like to be able to independently release different packages
+
+Below summarizes a timeline of releases for three packages, A, B and C, where B depends on A and C depends on B.
+
+
+![alt_text](20190305-modular-tensorflow/releases.png "Release plans")
+
+
+To summarize the above timeline:
+
+*   Different packages set their own release cadences
+*   Each package will set version boundaries for each of their dependencies.
+*   Each package is responsible for ensuring that all of their public APIs are working without any changes until the next major release
+*   Packages do not need to modify the minimum version requirements unless they start using newly introduced public API symbols.
+*   TF metapackage releases may choose to hold back individual packages in favor of faster releases. But dependency requirements have to be respected when doing so.
+*   Major releases still need to be coordinated.
+
+
+## Packaging Issue Scenarios
+
+
+We will now go over a few failure modes in the proposed environment, and propose how things need to be resolved. We will use a simple example, where we have three packages, A, B and C, where C depends on B and A, and B depends on A.
+
+
+#### Scenario 1: B does not work with a new release of A
+
+Potential issues:
+
+*   B uses non-public symbols from A:
+    *   B has to release a new version avoiding non-public APIs
+*   A changed public APIs:
+    *   A has to revert changes to its public APIs and release a patch version replacing the bad version
+
+
+#### Scenario 2: New release of B does not work with A
+
+Potential issues:
+
+*   B depends on unreleased APIs A exposed in a newer version
+    *   B needs to release a patch version, with just a change in the minimum required version of A
+
+
+#### Scenario 3: C and B depend on different minimum versions of A
+
+As both C and B have to define a range of versions they require from A, the max version should satisfy both constraints.
+
+
+#### Scenario 4: User installs C first, but then uninstalls A
+
+This is a pip issue. To help diagnose the problem, B and C should print that A is missing, and user needs to install that to use B or C.
+
+
```

Comment:
Who owns documentation for plugins? Is there plan for a shared site with plugin contribution?

Reply:
Each plugin resides in its own repository, so does the doc site (separated from TF core docs).

Reply:
CC @lamberta
It is actually a mix. There will have to be some plugins, like estimator we will declare as a part of core TF APIs, and document as such. For some other plugins, what @byronyi said will apply.